sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -686,6 +686,218 @@ case class EndsWith(left: Expression, right: Expression) extends StringPredicate
+    "otherwise returns a new string whose invalid UTF8 bytes sequences are replaced using the " +
+    "UNICODE replacement character U+FFFD.",
+  arguments = """
+    Arguments:
+      * str - a string expression
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_('Spark');
+       Spark
+      > SELECT _FUNC_(x'61');
+       a
+      > SELECT _FUNC_(x'80');
+       �
+      > SELECT _FUNC_(x'61C262');
+       a�b
+  """,
+  since = "4.0.0",
+  group = "string_funcs")
+// scalastyle:on
+case class MakeValidUTF8(srcExpr: Expression) extends UnaryExpression with ImplicitCastInputTypes
+  with NullIntolerant {
+
+  override def child: Expression = srcExpr
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringTypeAnyCollation)
+  override def dataType: DataType = child.dataType
+  override def nullable: Boolean = true
+
+  override def nullSafeEval(srcEval: Any): Any = {
+    srcEval.asInstanceOf[UTF8String].makeValid
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    defineCodeGen(ctx, ev, c => s"${ev.value} = $c.makeValid();")
+  }
+
+  override def prettyName: String = "make_valid_utf8"
+
+  override protected def withNewChildInternal(newChild: Expression): MakeValidUTF8 = {
+    copy(srcExpr = newChild)
+  }
+
+}
+
+/**
+ * A function that validates a UTF8 string, throwing an exception if the string is invalid.
+ */
+// scalastyle:off
+@ExpressionDescription(
+  usage = "_FUNC_(str) - Returns the original string if `str` is a valid UTF-8 string, " +
+    "otherwise throws an exception.",
+  arguments = """
+    Arguments:
+      * str - a string expression
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_('Spark');
+       Spark
+      > SELECT _FUNC_(x'61');
+       a
+  """,
+  since = "4.0.0",
+  group = "string_funcs")
+// scalastyle:on
+case class ValidateUTF8(srcExpr: Expression) extends UnaryExpression with ImplicitCastInputTypes
+  with NullIntolerant {
+
+  override def child: Expression = srcExpr
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringTypeAnyCollation)
+  override def dataType: DataType = child.dataType
+  override def nullable: Boolean = true
+
+  override def nullSafeEval(srcEval: Any): Any = {
+    val utf8string = srcEval.asInstanceOf[UTF8String]
+    if (!utf8string.isValid) throw QueryExecutionErrors.invalidUTF8StringError(utf8string)
+    else utf8string
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    nullSafeCodeGen(ctx, ev, c =>
+      s"""
+        if (!$c.isValid()) {
+          throw QueryExecutionErrors.invalidUTF8StringError($c);
+        } else {
+          ${ev.value} = $c;
+        }
+      """
+    )
+  }
+
+  override def prettyName: String = "validate_utf8"
+
+  override protected def withNewChildInternal(newChild: Expression): ValidateUTF8 = {
+    copy(srcExpr = newChild)
+  }
+
+}
+
+/**
+ * A function that tries to validate a UTF8 string, returning null if the string is invalid.
```

Comment:
```suggestion
 * A function that tries to validate a UTF8 string, returning NULL if the string is invalid.
```
Not sure whether "null" or "NULL" is more appropriate here. Can you check what we do in other expressions?